{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a4c0b26",
   "metadata": {},
   "outputs": [],
   "source": [
    "##PROCESSING OF 16S AMPLICON SEQUENCES##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c35f298",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Quality check\n",
    "for N in `ls zr*_R1.fastq`; do fastx_trimmer -f 10 -l 280 -z -i $N -o $N.gz; done\n",
    "for N in `ls zr*_R2.fastq`; do fastx_trimmer -f 7 -l 253 -z -i $N -o $N.gz; done\n",
    "\n",
    "for N in `ls p*_R1.fastq`; do fastx_trimmer -f 7 -z -i $N -o $N.gz; done\n",
    "for N in `ls p*_R2.fastq`; do fastx_trimmer -f 10 -l 240 -z -i $N -o $N.gz; done\n",
    "\n",
    "for N in `ls a*_R1.fastq`; do fastx_trimmer -f 7 -z -i $N -o $N.gz; done\n",
    "for N in `ls a*_R2.fastq`; do fastx_trimmer -f 10 -l 240 -z -i $N -o $N.gz; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0abd7cf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#A list of files to assembly is generated.\n",
    "ls *.fastq.gz | perl -pe 's/_R.*.fastq.gz//g' | sort | uniq >lista\n",
    "\n",
    "#Assembly using CASPER. #Run assembly_CASPER.sh script to generate works\n",
    "bash assembly_CASPER.sh mp_ensamble\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -j y\n",
    "#$ -S /bin/bash\n",
    "\n",
    "SEQS=$(pwd)\n",
    "SALIDAS=$(pwd)\n",
    "BIN=/usr/local/bin\n",
    "BIN2=/usr/local/bin\n",
    "BIN3=/usr/bin\n",
    "COUNT=0\n",
    "for FAA in `cat lista`\n",
    "do\n",
    "let COUNT=COUNT+1\n",
    "echo \"#!/bin/bash\" >$*.$COUNT.scr\n",
    "echo \"#$ -cwd\" >>$*.$COUNT.scr\n",
    "echo \"#$ -j y\" >>$*.$COUNT.scr\n",
    "echo \"#$ -S /bin/bash\" >>$*.$COUNT.scr\n",
    "echo \"$BIN/casper -d 12 -k11 -w 12 -g 0.6 <(zcat $SEQS/$FAA\"_R1.fastq.gz\") <(zcat $SEQS/$FAA\"_R2.fastq.gz\") -o $FAA\"_assembly\"\" >>$*.$COUNT.scr\n",
    "echo \"$BIN3/fastq_to_fasta -n  -i $FAA_assembly.fastq -o $FAA\"_assembly.fas\"\" >>$*.$COUNT.scr\n",
    "done\n",
    "\n",
    "\n",
    "#Running works in the cluster\n",
    "for N in `ls *.scr`; do qsub $N; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4740caac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Verifying sequence length for V3 region trimming\n",
    "~/anaconda3/bin/seqkit stats p*_assembly.fas \n",
    "~/anaconda3/bin/seqkit stats p*_assembly.fas "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06122eaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Removing the first 130 bases (V3 region) from the pulque samples \n",
    "for N in `ls p*_assembly.fas | sed 's/.fas//g'`; do fastx_trimmer -f 131 -i $N.fas -o $N.fasta; done\n",
    "for N in `ls a*_assembly.fas | sed 's/.fas//g'`; do fastx_trimmer -f 131 -i $N.fas -o $N.fasta; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4079074a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Verifying sequence length for pulque sequences\n",
    "~/anaconda3/bin/seqkit stats p*_assembly.fasta \n",
    "~/anaconda3/bin/seqkit stats p*_assembly.fasta \n",
    "#Renaming \n",
    "rm p*.fas a*.fas\n",
    "for N in `ls *.fas | sed 's/.fas//g'`; do mv $N.fas $N.fasta; done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c7e33c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Number of merged sequences per sample\n",
    "grep -c \"@M\" *_assembly.fastq > total_sec_ensambladas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aa8bb3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Sequences names are changed with “header.fasta.numbers.pl”. The script is runned for each sample\n",
    "for N in `ls *.fasta | sed 's/_assembly.fasta//g'`; do perl /home/angelica/scripts/header.fasta.numbers.pl $N $N\"_assembly.fasta\"; done\n",
    "#Concatenate all samples\n",
    "cat *.numbered.fas > mp_16S.fas\n",
    "#Edit sequence name to leave only the first part, which refers to sample name.\n",
    "perl -i.bak -pe \"s/\\ .*//g\" mp_16S.fas\n",
    "#Edit the name header\n",
    "sed -i 's/_/./' mp_16S.fas\n",
    "#Sequence clustering of OTUs at 97%  of identity was done with cd-hit-est, -c indicates the identity for clustering.\n",
    "cd-hit-est -i mp_16S.fas -c 0.97 -o mp_16S.fasout -T 20 -M 0\n",
    "#The clustering file is converted in a file that can be read by QIIME.\n",
    "perl -pne 's/\\t//g;s/^.*,//g;s/\\.\\.\\..*$//g;s/\\n/\\t/g;s/\\>Cluster\\ /\\n/g;s/\\>//g; eof && do{chomp; print \"$_ \\n\"; exit}' mp_16S.fasout.clstr > mp_16S.otu\n",
    "sed -i '1d' mp_16S.otu\n",
    "#Extract representative OTUs, -i indicates input file, -f is the fasta file with the sequences to be extracted. \n",
    "/home/qiime/bin/pick_rep_set.py -i mp_16S.otu -f mp_16S.fas -o mp_16S.rep.fna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46f6aaa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Removing singletons\n",
    "wc -l mp_16S.otu #236356\n",
    "awk 'NF < 3' mp_16S.otu | cut -f1 >ids_REMOVE_singletons.txt\n",
    "wc -l ids_REMOVE_singletons.txt #213787 ids_REMOVE_singletons.txt\n",
    "cut -f1 mp_16S.otu > rawids\n",
    "\n",
    "#Generating file with singleton's identities\n",
    "awk 'NR==FNR { values[$0]; next } !($0 in values)' ids_REMOVE_singletons.txt rawids  >repsetWOsingletons.txt\n",
    "\n",
    "\n",
    "#Extracting fasta without singletons \n",
    "perl -ne 'if(/^>(\\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV'repsetWOsingletons.txt mp_16S.rep.fna >lda_nosingletons.fna\n",
    "grep -c \">\" lda_nosingletons.fna #22569\n",
    "\n",
    "\n",
    "#Removing contaminants (no 16S sequences)\n",
    "parallel_assign_taxonomy_blast.py -i lda_nosingletons.fna -o ldaNO16S_screen1 -r /qiime/gg_otus-13_8-release/rep_set/70_otus.fasta -t /qiime/gg_otus-13_8-release/taxonomy/70_otu_taxonomy.txt -X LDA1\n",
    "\n",
    "#Identifying sequences with no hits on Greengenes db at 70% identity\n",
    "cat ldaNO16S_screen1/lda_nosingletons_tax_assignments.txt | grep \"No blast hit\" | cut -f1 >idsnohit.txt\n",
    "wc -l idsnohit.txt #2194 idsnohit.txt\n",
    "\n",
    "#Extracting sequences with no hit from fasta\n",
    "perl -ne 'if(/^>(\\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' idsnohit.txt lda_nosingletons.fna  >nohitGG.fasta\n",
    "\n",
    "#Verifying that the sequences don't have a hit in the last version of Silva db\n",
    "parallel_assign_taxonomy_blast.py -i nohitGG.fasta -o LDAsinhitGG -r /databases/silva/qiime/dada_silva138.fna -t /databases/silva/qiime/dada_silva138.tax.txt\n",
    "\n",
    "\n",
    "#Obtaining sequence ids without singletons and contaminants (screened 16S)\n",
    "grep -v \"No blast hit\" ldaNO16S_screen1/lda_nosingletons_tax_assignments.txt | cut -f1 >ggret1.txt\n",
    "grep -v \"No blast hit\" LDAsinhitGG/nohitGG_tax_assignments.txt | cut -f1 >ggret2.txt\n",
    "cat ggret*.txt >ret_screened_nosingleton.txt\n",
    "\n",
    "wc -l ret_screened_nosingleton.txt #22251 ret_screened_nosingleton.txt\n",
    "\n",
    "#Extracting sequences with no hit\n",
    "perl -ne 'if(/^>(\\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' ret_screened_nosingleton.txt  lda_nosingletons.fna >nosingleton_screened.fna\n",
    "\n",
    "#Aligning to identify chimeric sequences\n",
    "parallel_align_seqs_pynast.py -i nosingleton_screened.fna -o chimera_align -O 400 -X chimx\n",
    "\n",
    "#Searching chimeras\n",
    "parallel_identify_chimeric_seqs.py -m blast_fragments -i nosingleton_screened.fna -a chimera_align/nosingleton_screened_aligned.fasta -o singleton_chimera.txt -X chimerablast --id_to_taxonomy_fp /qiime/gg_otus-13_8-release/taxonomy/97_otu_taxonomy.txt -r /qiime/gg_otus-13_8-release/rep_set/97_otus.fasta -O 300\n",
    "wc -l singleton_chimera.txt #5289 singleton_chimera.txt\n",
    "\n",
    "#Obtaining chimeras ids\n",
    "awk '{print $1}' singleton_chimera.txt >chimeraids\n",
    "\n",
    "#Cleaning files\n",
    "awk 'NR==FNR { values[$0]; next } !($0 in values)' chimeraids ret_screened_nosingleton.txt >ids_nochimera_screened_nosingleton.txt\n",
    "\n",
    "wc -l  ids_nochimera_screened_nosingleton.txt #16962 ids_nochimera_screened_nosingleton.txt\n",
    "\n",
    "#Creating OTU table without chimeras, filtered, and without singletons\n",
    "awk 'NR==FNR { values[$0]; next } $1 in values { print }' ids_nochimera_screened_nosingleton.txt mp_16S.otu >nochimera_screened_singleton.otu\n",
    "\n",
    "#Creating new fasta file withou chimeras, filtered, and iwthout singletons\n",
    "perl -ne 'if(/^>(\\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV' ids_nochimera_screened_nosingleton.txt nosingleton_screened.fna >nochimera_screened_nosingleton.fna\n",
    "grep -c \">\" nochimera_screened_nosingleton.fna #16962\n",
    "\n",
    "\n",
    "#Asigning taxonmy using Silva db\n",
    "parallel_assign_taxonomy_blast.py -i nochimera_screened_nosingleton.fna -o tax -r /databases/silva/qiime/dada_silva138.fna -t /databases/silva/qiime/dada_silva138.tax.txt -O 300 -X tom\n",
    "\n",
    "#Cleaning taxonomy\n",
    "awk '{ for(i=1; i<=NF-2; i++) printf \"%s%s\", $i, (i<NF-2) ? OFS : ORS }' tax/nochimera_screened_nosingleton_tax_assignments.txt | perl -pe 's/\\ /\\_/g;s/_k/\\tk/;s/\\;_/\\t/g' >tax_clean.txt\n",
    "\n",
    "#Removing mithocondria and chloroplasts\n",
    "grep -i 'chloroplast' tax_clean.txt | cut -f1 >tmprm1\n",
    "grep -i 'mitochondria' tax_clean.txt | cut -f1 >>tmprm1\n",
    "\n",
    "#Creating new OTU table without chimeras, filtered, without singletons, mitochondria and chloroplasts\n",
    "awk 'NR==FNR { values[$0]; next } !($0 in values)' tmprm1 ids_nochimera_screened_nosingleton.txt >nochimera_screened_singleton_cpmt.ids\n",
    "awk 'NR==FNR { values[$0]; next } $1 in values { print }' nochimera_screened_singleton_cpmt.ids nochimera_screened_singleton.otu >nochimera_screened_singleton_cpmt.otu\n",
    "wc -l nochimera_screened_singleton_cpmt.otu #16888 nochimera_screened_singleton_cpmt.otu\n",
    "\n",
    "#Creating OTU FASTA file without chimeras, filtered, without singletons, mitochondria and chloroplasts\n",
    "perl -ne 'if(/^>(\\S+)/){$c=$i{$1}}$c?print:chomp;$i{$_}=1 if @ARGV'  nochimera_screened_singleton_cpmt.ids nochimera_screened_nosingleton.fna >nochimera_screened_singleton_cpmt.fna\n",
    "\n",
    "#Generating alignment with filtered OTUs\n",
    "ssu-prep -f nochimera_screened_singleton_cpmt.fna mezcaltax 40 prefixcluster.txt\n",
    "bash mezcaltax.ssu-align.sh\n",
    "ssu-mask --stk2afa mezcaltax\n",
    "\n",
    "#Cleaning alignment\n",
    "perl -i.bak -pe 's/\\./-/g;s/U/T/gi' mezcaltax/mezcaltax.bacteria.afa\n",
    "\n",
    "# The generated file has 16840 sequences instead of 1692. The difference is due to ssu-align identified Archaea and removed\n",
    "#Make phylogeny\n",
    "./fasttreeMP -nt -fastest ssu-align/mezcaltax/mezcaltax.bacteria.afa >mezcal.nwk\n",
    "\n",
    "\n",
    "#Creating OTU table in biom format\n",
    "make_otu_table.py -i nochimera_screened_singleton_cpmt.otu -o nochimera_screened_singleton_cpmt.biom\n",
    "biom convert --to-tsv -i nochimera_screened_singleton_cpmt.biom -o nochimera_screened_singleton_cpmt.biom.tsv\n",
    "ln -s tax_clean.txt nochimera_screened_singleton_cpmt.tax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a0ea79c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Phylogenetic tree \n",
    "cut -f1 mp_16S.biom.tsv>list_otu_ids\n",
    "/home/qiime/bin/filter_fasta.py -f chimalign/mp_16S.screened_aligned.fasta -o mp_16S.biom.fna -s list_otu_ids \n",
    "qsub -pe completenode 4 -N arbolote -b y -j y -cwd -V \" FastTreeMP -nt -gtr mp_16S.biom.fna > tree_mp_16S.nwk\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
